main/src/ca/uwaterloo/flix/language/phase/Typer.scala

```diff
@@ -1439,11 +1439,22 @@ object Typer {
           resultEff = valueEff
         } yield (valueConstrs, resultTyp, resultPur, resultEff)
 
-      case KindedAst.Expression.NewObject(clazz, _, loc) =>
-        val resultTyp = getFlixType(clazz)
-        val resultPur = Type.Impure
-        val resultEff = Type.Empty
-        liftM(List.empty, resultTyp, resultPur, resultEff)
+      case KindedAst.Expression.NewObject(clazz, methods, loc) => {
+        def visitJvmMethod(method: KindedAst.JvmMethod): InferMonad[(List[Ast.TypeConstraint], Type, Type, Type)] = method match {
+          case KindedAst.JvmMethod(ident, fparams, exp, returnTpe, pur, eff, loc) =>
+            for {
+              (constrs, bodyTpe, bodyPur, bodyEff) <- visitExp(exp)
+              _ <- expectTypeM(expected = returnTpe, actual = bodyTpe, exp.loc)
+            } yield (constrs, returnTpe, bodyPur, bodyEff)
+        }
+
+        for {
+          (constrs, _, _, _) <- seqM(methods map visitJvmMethod).map(unzip4)
+          resultTyp = getFlixType(clazz)
+          resultPur = Type.Impure
+          resultEff = Type.Empty
```

Comment:
We can discuss later whether object creation is actually impure or not.

Reply:
Surely it has to be impure? We're creating a new object, which has a different identity. [This test](https://github.com/flix/flix/blob/master/main/test/flix/Test.Exp.Jvm.NewObject.flix#L32) demonstrates that it has to be impure, yes?

Reply:
Yes. Ok you are right.